gradle-jdks-setup/README.md

```diff
@@ -0,0 +1,245 @@
+  - is rendered from `JdksExtension#caCerts`
+- `gradle/jdks`
+  - contains a list of directories in the format `<jdk_major_version>/<os>/<arch>` that contain 2 files: 
+    - `download-url` full url path for the jdk, os and arch. Rendered from `JdksExtension#jdks` configured in step 2
+    - `local-path` the local name of the file. Rendered based on the distribution-name, version and the [hash](../gradle-jdks/src/main/java/com/palantir/gradle/jdks/JdkSpec.java) 
+  - it generates all the JDK versions configured in [JdksExtension](../gradle-jdks/src/main/java/com/palantir/gradle/jdks/JdksExtension.java)
+
+Running the patched `./gradlew` script will add extra configurations required for Intelij:
+```
+project-root/
+├── .gradle/
+│   ├── config.properties
+```
+
+- `.gradle/config.properties` - sets up `java.home` to the gradle JDK daemon path.
+
+## How it works ?
+There are 2 main entry points for running Gradle. Both of these would need to support installing/using the specified JDK. 
+
+* `./gradlew(.bat)` script
+* Running a Gradle build from inside `Intellij`
+
+### Supporting Gradle JDK auto-management from `./gradlew(.bat)` scripts
+
+The `./gradlew` script is the main entry point for running Gradle builds. It is a shell script that downloads the Gradle wrapper jar and runs it with a JDK. The script is generated by the `gradle-wrapper` plugin and is stored in the `gradle/wrapper` directory of the project.
+We are modifying the `./gradlew` script to use the JDK setup by patching it using the scripts: [gradlew-patch.sh](../gradle-jdks/src/main/resources/gradlew-patch.sh) and [gradle-jdks-setup.sh](src/main/resources/gradle-jdks-setup.sh). 
+The patching is done by the [`wrapperJdkPatcher` task.](../gradle-jdks/src/main/java/com/palantir/gradle/jdks/GradleWrapperPatcher.java).
+The patch script does the following: 
+* downloads all the JDKs that are configured in the `gradle/jdks` directory [see above the dirctory structure](#gradle-jdk-configuration-directory-structure) 
+* delegates to `gradle-jdks-setup.jar` ([setup class](src/main/java/com/palantir/gradle/jdks/setup/GradleJdkInstallationSetup.java)) the installation of the JDKS and the certs (configured in `gradle/certs`)
+* sets the gradle property `org.gradle.java.home` to the installation path of the JDK configured in `gradle/gradle-daemon-jdk-version`. Hence, `./gradlew` will retrieve this java installation and it will run the wrapper using this java installation.
+
+
+### Running a Gradle build from inside `Intellij`
+`Intelij` doesn't use the `./gradlew` script, instead it uses the [Gradle Tooling API](https://docs.gradle.org/current/userguide/third_party_integration.html#sec:embedding_introduction).
+In the Intellij plugin `palantir-gradle-jdks` an ExternalSystemTaskNotificationListener will be registered that for every `Gradle` task of type `RESOLVE_PROJECT` or `EXECUTE_TASK` will:
```

Comment:
I wouldn't go into detail about the internals (likely to go out of date eventually, people can just read the code if they want to deeply understand it), just say that before and Gradle task or refresh is run: